Switch internal usage to ID-based triples

mberr opened the pull request:
This PR partially implements the changes from #57 without strictly separating labelling and triples (yet).

In particular:
* it uses ID-based triples wherever possible, including inverse/duplicate relation detection
* it postpones inverse relation creation to the creation of training instances (fixes #146, fixes #148)
* it updates the leakage workflow to work ID-based. While doing so I recognized that the reproduction of FB15k-237 is currently not working. Since it isn't on master as well, it will not be fixed in this PR, but postponed to a later PR.

mberr committed:
Do not store labeled triples in TriplesFactory

mberr committed:
Add TODO

mberr committed:
Start converting leakage detection to IDs

mberr committed:
Update type annotation to use generic

mberr committed:
Add more TODO

mberr committed:
Extract utility method to label triples

mberr committed:
"Update" TriplesFactory.split to use ID-based triples

mberr committed:
Update get_most_frequent_relations to return relation IDs

mberr committed:
Update get_triples_for_relations

mberr committed:
Update new_with_relations

mberr committed:
Update new_without_relations

mberr committed:
Update new_with_restriction

mberr committed:
Create inverse mappings and all vectorized mappings in __post_init__ hook

mberr committed:
Update label_triples to use precomputed vectorized labelers

mberr committed:
Update word clouds to more efficiently create labeled entity pairs /relations

mberr committed:
Improve efficiency of word-cloud by pre-filtering ID-based

mberr committed:
Relax callable type annotation

mberr committed:
Re-use vectorized labelers

mberr committed:
Start updating TriplesFactory.split

mberr committed:
Update comment

mberr committed:
Update _tf_cleanup_deterministic

mberr committed:
Update _prepare_cleanup

mberr committed:
Make max_ids optional

mberr committed:
Update to use torch random states

mberr committed:
Add TODO

mberr committed:
Move utility to utils

mberr committed:
Add unittest for torch_is_in_1d

mberr committed:
Remove todo; update docstring

mberr committed:
Remove parameter

mberr committed:
Add TODOs

mberr committed:
Fix type annotation

PyKEEN-bot committed:
Trigger CI

mberr committed:
Merge remote-tracking branch 'origin/internally-only-use-id-based-triples' into internally-only-use-id-based-triples

mberr committed:
Add documentation to get_absolute_split_sizes

mberr committed:
Reformat

mberr committed:
Add test for get_absolute_split_sizes and fix the method

mberr committed:
Add documentation

mberr committed:
Add test for normalize_ratios

mberr committed:
Remove obsolete TODO

mberr committed:
Add additional test

mberr committed:
Add epsilon for floating point comparison

cthoyt commented on src/pykeen/triples/triples_factory.py:
```diff
@@ -745,62 +895,76 @@ def _tf_cleanup_all(
             reference, other = _tf_cleanup_deterministic(reference, other)
         rv.append(other)
-    return [reference, *rv]
+    return reference, *rv
```
this fails in py37, so you have to keep the brackets

mberr replied:
5215a14

mberr committed:
Revert change to maintain 3.7 compatibility

cthoyt committed:
Pass flake8 and mypy

Trigger CI

cthoyt committed:
Remove unused functions

cthoyt commented on src/pykeen/triples/triples_factory.py:
```diff
@@ -540,45 +651,84 @@ def get_most_frequent_relations(self, n: Union[int, float]) -> Set[str]:
-        return (
-            np.isin(self.triples[:, 0], entities, invert=invert)
-            & np.isin(self.triples[:, 2], entities, invert=invert)
+    def entities_to_ids(self, entities: Collection[Union[int, str]]) -> Collection[int]:
```
Might be better to type as `Union[Collection[int], Collection[str]]`. The implementation is agnostic, but I think this gives a better idea what's going on from a documentation standpoint

cthoyt replied:
Same for all of the other functions using this one and the equivalent relation one

mberr replied:
The former one allows mixed types, so this would restrict the API. However, I am not sure whether there is a use case for providing mixed ones :sweat_smile:

cthoyt commented on src/pykeen/triples/triples_factory.py:
```diff
@@ -540,45 +651,84 @@ def get_most_frequent_relations(self, n: Union[int, float]) -> Set[str]:
+        )
 
-    def get_triples_for_relations(self, relations: Collection[str], invert: bool = False) -> LabeledTriples:
+    def get_triples_for_relations(self, relations: Collection[Union[int, str]], invert: bool = False) -> LabeledTriples:
```
Apparently not, since this is only used for a test

mberr committed:
Change return type of get_triples_for_relations from LabeledTriples to MappedTriples

cthoyt commented on tests/test_triples_factory.py:
```diff
@@ -397,3 +398,48 @@ def test_inverse_triples(self):
+
+        # check consistency with ratios
+        rel_size = np.asarray(sizes) / n_total
+        decimal = np.floor(-np.log10(1 / n_total))
```
are you doing `-np.log10(1 / n_total)` to check for numeric stability? otherwise you could directly do `np.log10(n_total)`

mberr replied:
96d745b  and 371add8

mberr committed:
Simplify code

mberr committed:
Add comment

cthoyt committed:
Merge branch 'master' into internally-only-use-id-based-triples

mberr committed:
Fix typo

mberr committed:
Add base case

mberr committed:
Fixes in _prepare_cleanup

mberr committed:
Fixes in test data type

mberr committed:
Fix concat along wrong dimension

mberr committed:
Fix comment

mberr committed:
Fix randomized cleanup

mberr committed:
Fix TriplesNumericLiteralsFactory.__init__

mberr committed:
Fix test_count_inverse_frequencies

mberr committed:
Fix test_find_leak_assymetric

mberr committed:
Extract utility

mberr committed:
More restrictive type annotations

mberr committed:
Pure index-based inverse triples

mberr committed:
Revert creating inverse triples only when creating training instances

mberr committed:
Add TODO

mberr committed:
Explicitly copy relation_to_inverse when splitting/restricting

mberr committed:
Fix ID-based inverse triples creation

mberr committed:
Fix labels of inverse relations in new_relation_to_id creation

mberr committed:
Fix carry over flag

mberr committed:
Add comment to failing unittest

mberr commented on tests/test_triples_factory.py:
```diff
@@ -86,6 +87,9 @@ def test_automatic_inverse_detection(self):
         """Test if the TriplesFactory detects that the triples contain incomplete inverses and creates correct ids."""
+        # comment(mberr): from my pov this behaviour is faulty: the triples factory is expected to say it contains
+        # inverse relations, although the triples contained in it are not the same we would have when removing the
+        # first triple, and passing create_inverse_triples=True.
```
@cthoyt @lvermue 
The behaviour was changed in this PR. If you approve these changes, I'll adapt the unittest.

cthoyt replied:
You're probably right - using the `_inverse` suffix to assume that there were already inverses was a bit sketchy. I'm okay with changing this behavior - thanks for flagging it for specific discussion

mberr replied:
Right now, I still run the check. If I find something looking like an inverse, I emit a warning, and discard all triples with these relations. They get re-created automatically when creating the ID-based training instances.

cthoyt replied:
I think that's a really good compromise

mberr replied:
I agree however that this still is sketchy since it relies on a very specific naming scheme. For content-based inverse relation detection we already have the sealant. Running its detection whenever we create a triples factory might however slow things down.

We could add a flag to manually select to trigger a check, but then again, the user could just directly use the sealant.

lvermue replied:
Shouldn't we delete it completely then?

cthoyt replied:
Let's take care of that in the future PR where we make the functionality work

mberr replied:
I am also in favour of deleting it.

cthoyt replied:
Okay I didn’t have a strong opinion, so please @mberr delete the deprecated code 💯

cthoyt commented on tests/test_triples_factory.py:
```diff
@@ -251,45 +255,45 @@ def test_split_multi(self):
         new_testing, new_testing = _tf_cleanup_all([training, testing])
-        self.assertEqual(expected_training, new_training.tolist())
-        self.assertEqual(expected_testing, new_testing.tolist())
+        assert (expected_training == new_training).all()
```
you can use `self.assertTrue`. The only reason I've switched a lot of assert statements into these functional forms is that unittest/pytest will give more information/context on failures

mberr replied:
For me it is exactly the opposite:
With self.assert*, I get a fancy comparison between expected and the actual value, but I cannot directly see the stack trace of the failed comparison. With "assert ..." I get a full stack trace of all participating methods. Maybe that is due to having `pytest` as default test runner?

cthoyt replied:
oh boy... Are you checking the tests using tox or an IDE? Tox's configuration uses pytest, but I think typically I use unittest in the IDE becuase it integrates better.

mberr replied:
IDE, Pycharm in particular. It actually might come from PyCharm magic things.

That's how they look like

```ptb
FAILED [ 14%]
 
{'a': 0, 'a.': 1, 'a._inverse': 2} != {'a': 0, 'a.': 2, 'a._inverse': 3, 'a_inverse': 1}

Expected :{'a': 0, 'a.': 2, 'a._inverse': 3, 'a_inverse': 1}
Actual   :{'a': 0, 'a.': 1, 'a._inverse': 2}
<Click to see difference>

self = <tests.test_triples_factory.TestTriplesFactory testMethod=test_automatic_incomplete_inverse_detection>
first = {'a': 0, 'a.': 2, 'a._inverse': 3, 'a_inverse': 1}
second = {'a': 0, 'a.': 1, 'a._inverse': 2}, msg = None

    def _patched_equals(self, first, second, msg=None):
        try:
>           old(self, first, second, msg)

/snap/pycharm-professional/226/plugins/python/helpers/pycharm/teamcity/diff_tools.py:32: 
```

lvermue replied:
I agree with @mberr. I use the pytest integration in Pycharm as well and `assert` gives a much more sensible stack trace of the failed comparison. It actually shows exactly where the error lies in the code no matter how many inheritances or function calls one is away from the test call itself. `self.assertTrue` only mentions the failed test, but only shows a stack trace to where the comparison function itself failed (not where the result actually comes from), which is interesting but not informative at all 🤷‍♂️

lvermue replied:
@cthoyt Do we actually have any documentation on our "testing strategy" and with which design considerations our tests were setup?

cthoyt replied:
No, there isn't a place documenting the testing style. @lvermue I'm surprised you're having the same problem as Max ☹️

mberr committed:
create inverse triples on demand

mberr committed:
Remove obsolete parameter

cthoyt commented on src/pykeen/triples/triples_factory.py:
```diff
@@ -456,6 +422,29 @@ def from_path(
             compact_id=compact_id,
         )
 
+    def with_other_triples(
```
this is no doubt a good function, but needs a better name ;)

mberr replied:
Suggestions? I first had `replace_triples` but that sounded like an in-place operation.

mberr committed:
Fix unittest

mberr committed:
Add max_relation_id to TriplesFactory

mberr committed:
Swap properties for backward compatibility

mberr committed:
Fix triple filtering

mberr committed:
Fix TriplesNumericLiteralsFactory

mberr committed:
Adjust test

mberr committed:
Increase logging

mberr committed:
Adjust tests

mberr commented on src/pykeen/triples/triples_factory.py:
```diff
@@ -42,6 +43,7 @@ def _create_multi_label_tails_instance(
     use_tqdm: Optional[bool] = None,
 ) -> Dict[Tuple[int, int], List[int]]:
     """Create for each (h,r) pair the multi tail label."""
+    # TODO: Replace by usage of torch.unique(return_inverses=True)?
```
cf. #197

mberr committed:
Remove unused method

mberr committed:
Re-introduce check

mberr committed:
Rename method

mberr committed:
Inline method used only in test

mberr committed:
Unify usage of triples restriction

cthoyt commented on src/pykeen/triples/triples_factory.py:
```diff
@@ -730,7 +706,7 @@ def new_with_relations(
             f' and {mask.sum()}/{self.num_triples} triples in {self}',
         )
-        return self.with_other_triples(mapped_triples=self.mapped_triples[mask])
+        return self.clone_and_exchange_triples(mapped_triples=self.mapped_triples[mask])
```
since almost all of the calls with this are using a mask, maybe you should even make a wrapper function that does this
```python
def with_mask(self, mask):
    return self.clone_and_exchange_triples(mapped_triples=self.mapped_triples[mask])
```

mberr replied:
Now it is only used twice: once with mask and once without :sweat_smile:

cthoyt commented on tests/test_leakage.py:
```diff
@@ -116,7 +116,9 @@ def test_find_leak_assymetric(self):
         # Test looking up inverse triples
-        test_leaked = test_factory.get_triples_for_relations(sealant.inverse_relations_to_delete)
+        test_leaked = test_factory.mapped_triples[
+            test_factory.get_mask_for_relations(relations=sealant.inverse_relations_to_delete, invert=False)
```
this is now the only place TriplesFactory.get_mask_for_relations is being used, right? maybe not necessary to test anymore?

mberr replied:
It is also used in `new_with_restriction`

mberr committed:
Update reindex to directly use IDs

mberr committed:
Fix re-index

mberr committed:
Fix _generate_vectorized_lookup

mberr committed:
Fix reindex

mberr committed:
Merge remote-tracking branch 'origin/master' into internally-only-use-id-based-triples

mberr committed:
Fix tests

mberr committed:
Compute candidate inverse and duplicate relations at once

mberr committed:
Improve logging of new_with_restriction

mberr committed:
Pass create_inverse_triples to validation and test if requested

cthoyt commented on src/pykeen/triples/leakage.py:
```diff
@@ -65,29 +65,49 @@ def __init__(
+        relations_sorted = sorted(relations.keys())
+        n_relations = len(relations_sorted)
+        size = numpy.asarray([len(relations[r]) for r in relations_sorted])
+        # we are not interested in self-similarity, thus we set it to zero
```
well technically self similarity means that the relation is reflexive. that's sometimes interesting to know

mberr replied:
This would be for r and r^-1, right?

mberr replied:
b97016e computes also this similarity

mberr committed:
some code refactoring for higher readability

mberr committed:
Extract method again

mberr committed:
Add method for pruning search space based on upper bound

mberr committed:
Extract relations to keep selection

mberr committed:
fix

mberr committed:
Merge remote-tracking branch 'origin/master' into internally-only-use-id-based-triples

# Conflicts:
#	src/pykeen/triples/triples_factory.py

mberr committed:
Fix merge artifact

mberr committed:
Add missing return type annotation

mberr committed:
Fix type annotation

mberr committed:
Add docstring to _generate_compact_vectorized_lookup

mberr committed:
Add comment

mberr committed:
Add docstring

mberr committed:
Add assertion

mberr committed:
Add tests for _generate_compact_vectorized_lookup

mberr committed:
ensure that generated test id tensor does not cover all numbers

mberr committed:
Add test for _translate_triples

mberr committed:
Extract method

mberr committed:
simplify code

cthoyt committed:
Reduce complexity

it turns out that before it was O(n^2) and this is O(n)... not that it makes a big difference on this scale but still

cthoyt committed:
Loop is so fast that tqdm isn't necessary

cthoyt committed:
Update logging

mberr committed:
Remove redundant line

mberr committed:
cast to python int

mberr committed:
Merge remote-tracking branch 'origin/internally-only-use-id-based-triples' into internally-only-use-id-based-triples

mberr committed:
Remove obsolete comment

mberr committed:
Extract common formatting for relative comparison

mberr committed:
try to simplify code for better readability

mberr committed:
Reformatting

mberr committed:
Extract method for candidate generation

mberr committed:
Disable pre-filtering for debugging

mberr committed:
Fix jaccard thresholding

mberr committed:
Yet another way to compute Jaccard similarity

mberr committed:
Rename method

mberr committed:
remove unused methods

mberr committed:
Fix __all__

mberr committed:
Improve documentation

mberr committed:
Add more documentation

mberr committed:
Improve doc

mberr committed:
Extract common code

mberr committed:
Add test for _relations_to_sparse_matrices

mberr committed:
Add test for _jaccard_similarity_scipy

cthoyt committed:
Reformat code

cthoyt committed:
Split function further preparing for testing

cthoyt committed:
Add test block with TODOs for get_candidate_pairs

@mberr you're up :)

mberr committed:
Simplify method

mberr committed:
Fix counting entity pairs

mberr committed:
Fix onehottiness

mberr committed:
Fix onehottiness

mberr committed:
Add test

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Punt this test for later

cthoyt committed:
Pass flake8

Trigger CI

cthoyt committed:
Merge branch 'master' into internally-only-use-id-based-triples

cthoyt committed:
Update triple generation to use indexes better

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Update generation.py

Trigger CI

cthoyt committed:
Remove deprecated code

Trigger CI

mberr committed:
Use as_tensor instead of tensor

mberr committed:
Re-use utility method

mberr committed:
use generic type annotation

mberr committed:
Add TorchRandomHint to typing.__all__

mberr committed:
Add ensure_torch_random_state to utils.__all__

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Merge branch 'master' into internally-only-use-id-based-triples

PyKEEN-bot committed:
Trigger CI

cthoyt commented on src/pykeen/triples/generation.py:
```diff
@@ -43,10 +45,13 @@ def generate_triples(
-            [new_entity_id[h], new_relation_id[r], new_entity_id[t]]
-            for h, r, t in rv
-        ], dtype=int)
+        rv = torch.as_tensor(
```
@mberr just curious - what's the difference between creating the tensor directly and using as_tensor?

mberr replied:
According to the doc of [torch.tensor](https://pytorch.org/docs/stable/generated/torch.tensor.html) and [torch.as_tensor](https://pytorch.org/docs/stable/generated/torch.as_tensor.html), `torch.as_tensor` is a more clever version which avoids creating copies when possible.

Also, at least in 1.6 `torch.tensor` did not have a proper type annotation which resulted in pylint warnings of `torch.tensor` not being callable, whereas `torch.as_tensor` didn't have this. This issue may have been fixed until then though.

cthoyt reviewed (approved):
🥳 🦜
